### ISSUE TITLE
Introduce catalogs support

### ISSUE THREAD

**Issue by jakeleventhal** [oven-sh/bun#4844]
### What is the problem this feature would solve?

In a monorepo, it would be cool to specify a single set of package pins at the repo root and those could then just be used throughout the repository instead of needing to be specified in numerous places.

Here is the PNPM RFC for accomplishing this: https://github.com/pnpm/rfcs/pull/1

### What is the feature you are proposing to solve the problem?

Sharable package pins accross a monorepo

**Comment by espenja** [comment]
Agree with this. It feels very unnecessary being forced to install "common" packages for a monorepo into every single workspace package. Examples include specific typescript version, eslint, prettier. Packages installed in the root dev/dependencies should be automatically resolvable by all workspace packages. If a package _must_ use a specific version it can be defined in it's respective package.json. This should further lead to individual packages not straying away from the versions used by the entire repo.

**Comment by Hebilicious** [comment]
Isn't this resolutions / overrides support ? #1134

**Comment by jakeleventhal** [comment]
@Hebilicious no, this is different

This PR is really for monorepos. For instance, if I use axios, i want to be able to specify that axios should use version `1.2.3`. If my monorepo has 30 package.json files that use axios, i want to just be able to specify `1.2.3` **once** in some root configuration and then all my packages can just use `"axios": "workspace:*"`

**Comment by Hebilicious** [comment]
> @Hebilicious no, this is different
> 
> This PR is really for monorepos. For instance, if I use axios, i want to be able to specify that axios should use version `1.2.3`. If my monorepo has 30 package.json files that use axios, i want to just be able to specify `1.2.3` **once** in some root configuration and then all my packages can just use `"axios": "workspace:*"`

Bun could add `workspace:*` support to resolutions/overrides so these would get resolved without introducing a new configuration syntax. You'd specify `"axios": "1.2.3:*"` in the `resolutions`/`overrides` of a monorepo root package.json.

**Comment by jakeleventhal** [comment]
@Hebilicious and how would that work for child package.json? what version specifier would you use for `axios`

**Comment by Hebilicious** [comment]
> @Hebilicious and how would that work for child package.json? what version specifier would you use for `axios`

You can use anything, incuding `workspace:*`. It will get replaced by whatever version overrides/resolutions has.

**Comment by jakeleventhal** [comment]
Solution worksforme

But I think consulting the PNPM rfc makes should be done. a lot of thought/effort went into that

**Comment by Hebilicious** [comment]
> Solution worksforme
> 
> But I think consulting the PNPM rfc makes should be done. a lot of thought/effort went into that

This RFC is quite long, I would suggest you update the description here with more accurate information. There's a subtle difference with overrides/resolutions that is documented [here](https://github.com/pnpm/rfcs/blob/main/text/0001-catalogs.md#comparison-to-overridesresolutions).
Note that this isn't implemented by any package manager so far, so it would be a non standard behaviour that bun could align with pnpms if they decide to go with `catalog`

**Comment by jakeleventhal** [comment]
I don't really have much skin in the game. Either solution works, but I think for the JS ecosystem it would be preferable if PNPM and bun were in sync here. That way, switching to bun can be even more of a "drop in replacement"

**Comment by unional** [comment]
> > @Hebilicious and how would that work for child package.json? what version specifier would you use for `axios`
> 
> You can use anything, incuding `workspace:*`. It will get replaced by whatever version overrides/resolutions has.

I would suggest to recommend only `workspace:*` (or `^`. Personally I think `workspace:^1.2.3` is noise and can complicate things quite a lot).

Replacing "anything" is technically correct as it is really just an override,
but using `workspace:*` have specific meaning.

**Comment by jakeleventhal** [comment]
Update, this is now implemented in `pnpm@9.5.0` (see https://github.com/pnpm/pnpm/issues/7072)

@Jarred-Sumner should bun just "work" here with the same syntax? Hard to tell with the binary bun lockfile

**Comment by chimericdream** [comment]
I'm curious if there is any update on whether or when this might be added. We have had a number of minor bugs recently where we had slight mismatches in dependency versions (e.g. i18next `^23.10.1` vs `23.12.2`), resulting in two copies of the affected libraries ending up in our bundled app.

For now, I am planning to pin everything to a particular version across our whole monorepo, but that isn't a sustainable approach.

**Comment by austinm911** [comment]
Since it's not mentioned, here are the pnpm docs for [pnpm catalogs](https://pnpm.io/catalogs)

**Comment by coolcorexix** [comment]
I tried set workspace:* and it just does not work because this is only used to mention the sub-repo of monorepo not the used packages
reference: https://bun.sh/docs/install/workspaces

**Comment by ohroy** [comment]
any news?

**Comment by necmettindev** [comment]
bump for bun team

**Comment by mi4uu** [comment]
bump! we need this

**Comment by karyanayandi** [comment]
bump! this feature very helpful

**Comment by bbarks15** [comment]
bump! only thing stopping me moving away from `pnpm`

**Comment by lucashfreitas** [comment]
bump! only thing holding us from moving away from pnpm

**Comment by mrtnbroder** [comment]
This feature is also stopping me rn from moving to bun (from pnpm)! I'd gladly help if there is something I can do!

**Comment by erickreutz** [comment]
We need this also

**Comment by ibufu** [comment]
+1

**Comment by versecafe** [comment]
Implementation proposal from #18939 

`package.json` or `bunfig.toml` based implementation of [pnpm catalogs](https://pnpm.io/catalogs)

root json
```json
{
  "workspaces": [ "packages/**", "apps/**" ],
  "catalogs": [ { "drizzle-orm": "0.41.0" } ],
}
```

or

```toml
# bunfig.toml
[test]
preload = "./happydom.ts"

[catalogs]
"drizzle-orm" = "0.41.0"
```

with end use looking like

```json
{
  "dependencies": {
    "drizzle-orm": "catalog:",
    "next": "^15",
  }
}
```

**Comment by Marius-brt** [comment]
Personally I think all the gestion of the packages for this should stay in the root `package.json`. I think an implementation like this should be the best considering that workspaces are already declared there.

Example of root `package.json` :
```json
{
  "name": "package-name",
  "version": "0.0.1",
  "scripts": {
      ...
  },
  "devDependencies": {
      ...
  },
  "workspaces": ["apps/*", "packages/*"],
  "catalogDependencies": {
      "typescript": "5.7.3",
      "@types/bun": "^1.2.9"
  }
}
```

For me this feature is stopping me in the usage of Bun in monorepo

**Comment by kravetsone** [comment]
The issue is interesting and I totally agree that it should be defined in the package.json

But what do we do with the JSON schema (but it is not strict as I remember)?
What if npm wants to add something with the same key...

**Comment by Marius-brt** [comment]
Yes I know that a problem. But if Bun implement the `pnpm-worspace.yml` it will be weird. The best would be that a common method would be follow by pnpm and bun but also other package manager.

**Comment by jakeleventhal** [comment]
> Yes I know that a problem. But if Bun implement the `pnpm-worspace.yml` it will be weird. The best would be that a common method would be follow by pnpm and bun but also other package manager.

it should be a universal implementation that respects pnpm-worksapce and package.json so that bun is plug and play

**Comment by kravetsone** [comment]
> Yes I know that a problem. But if Bun implement the `pnpm-worspace.yml` it will be weird. The best would be that a common method would be follow by pnpm and bun but also other package manager.

bun should not implement it yeah
maybe only migration can help

lets wait what Bun team will say

**Comment by Jarred-Sumner** [comment]
We are planning to implement support for catalogs. 

We probably will later add a pnpm-workspace.yml migration path, but unlikely with the initial version of catalogs.

**Comment by dylan-conway** [comment]
`bun install` will ship support for catalogs in the next release (Bun v.1.2.24). Catalogs work the same way as in pnpm, except they are listed under `"workspaces.catalog(s)"` in your root package.json instead of in `pnpm-workspaces.yaml`. Example:

`./package.json`:
```json
{
  "name": "bun-catalogs",
  "version": "1.0.0",
  "workspaces": {
    "packages": ["packages/*"],
    "catalog": {
      "jquery": "3.7.1"
    },
    "catalogs": {
      "beta": {
        "zod": "4.0.0-beta.20250505T195954"
      }
    }
  }
}
```

`./packages/pkg1/package.json`:
```json
{
    "name": "pkg1",
    "dependencies": {
        "jquery": "catalog:",
        "zod": "catalog:beta"
    }
}
```

`bun publish` will also replace `catalog:` with the catalog version, similar to `workspace:`.

**Comment by evelant** [comment]
Awesome! Great to see bun getting more modern package management abilities. Are there plans to add more control over hoisting (a more "correct" node_modules like pnpm does)? Lots of discussion over in https://github.com/oven-sh/bun/issues/533
